src/simp.c

```diff
@@ -927,20 +942,30 @@ static tree_t simp_case(tree_t t, simp_ctx_t *ctx)
       const int nassocs = tree_assocs(alt);
       for (int j = 0; j < nassocs; j++) {
          tree_t a = tree_assoc(alt, j);
-         if (tree_subkind(a) == A_NAMED) {
-            tree_t n = tree_name(a);
-            switch (tree_kind(n)) {
-            case T_LITERAL:
-            case T_STRING:
-            case T_REF:
-               break;
-            default:
-               if (eval_possible(n, ctx->registry, ctx->mir)) {
-                  n = eval_try_fold(ctx->jit, n, ctx->registry, NULL, NULL);
-                  tree_set_name(a, n);
-               }
-               break;
+         switch (tree_subkind(a)) {
+         case A_NAMED:
+            {
+               tree_t n = tree_name(a);
+               tree_t f = simp_case_choice(n, ctx);
+               if (n != f)
+                  tree_set_name(a, f);
+            }
+            break;
+
+         case A_RANGE:
+            {
+               assert(tree_ranges(a) == 1);
+               tree_t range = tree_range(a, 0);
+               tree_t l = tree_left(range);
+               tree_t lf = simp_case_choice(l, ctx);
+               if (l != lf)
+                  tree_set_left(range, lf);
+               tree_t r = tree_right(range);
+               tree_t rf = simp_case_choice(r, ctx);
+               if (r != rf)
+                  tree_set_right(range, rf);
```

Comment:
```suggestion
               tree_set_right(range, simp_case_choice(r, ctx));
```

(and elsewhere)

There's no harm in setting it to the same value and looks less noisy.

Reply:
I'd agree and that was actually how I initially did it. But with that I get several crashes of `Fatal: Write to object in frozen arena`. Is there something else I need to account for?

Reply:
I guess that happens because `simp_case` is called from `simp_tree_global` as well `simp_tree_local` (I can't remember why) and the `T_ALTERNATIVE` won't necessarily have a writable copy during elaboration.

I think you could probably fix that by moving this logic to a new `simp_case_alternative()` that's called from `simp_tree_local` only for `T_ALTERNATIVE`.